File: 0044-Wildcard-Matching/lc0044.py
# Python solution.

# Finite-state machine method (Discuss solution)
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        transfer = {}
        state = 0

        # construct state transfer
        for c in p:
            if c == '*':
                transfer[state, c] = state
            else:
                transfer[state, c] = state+1
                state += 1

        # searching all possible states
        acc_state = state
        states = set([0])
        for c in s:
            states = set([transfer.get((state, token)) for state in states for token in [c, '*', '?']])

        return acc_state in states

File: 0044-Wildcard-Matching/lc0044.js
// JavaScript solution.

/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    let i = 0, j = 0;
    let sm = 0, star = -1;
    let m = s.length, n = p.length;
    while (i < m) {
        if (j < n && (s[i] === p[j] || p[j] === '?')) {
            i++;
            j++;
        } else if (j < n && p[j] === '*') {
            star = j++;
            sm = i;
        } else if (star >= 0) {
            j = star+1;
            i = ++sm;
        } else { return false; }
    }
    while (j < n && p[j] === '*') { j++; }
    return j === n;
};
